### PR TITLE
Major refactor of Metal header generator

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Creates a device abstraction so that each compat-string we support gets described separately.

Regression tests pass with the headers generated by the script.

### CODE REVIEW

File: metal_header/device.c++ [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+void Device::emit_def_value(std::string name, const node &n, std::string field) {
+  uint32_t irqs;
+  os << "#define __METAL_" << n.handle_cap() << "_INTERRUPTS \t\t";
+  if (name.compare("interrupts") == 0) {
+    irqs = n.get_fields_count<uint32_t>(name);
+  } else {
+    irqs = n.get_fields_count<std::tuple<node, uint32_t>>(name);
+  }
+  os << irqs << "\n";
+  os << "#define METAL_MAX_" << field << "\t __METAL_" << n.handle_cap() << "_INTERRUPTS\n\n";
```
Comment (bsousi5) [inline metal_header/device.c++]: I think this method, should not emit the METAL_MAX_XXX but instead accumulated the interrupts, per field type.
So would need a switch statement for the field arg, and separate variable for each field type.

Then have a separate virtual method can emit the META_MAX_XXX with the accumulated value.

File: metal_header/freedom-metal_header-generator.c++ [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+  for(auto it = devices.begin(); it != devices.end(); it++) {
+    (*it)->create_machine_macros();
+  }
+  os << std::endl << "#else /* ! __METAL_MACHINE_MACROS */" << std::endl << std::endl;
+
+  for(auto it = devices.begin(); it != devices.end(); it++) {
+    (*it)->create_defines();
+  }
+  os << std::endl;
```
Comment (bsousi5) [inline metal_header/freedom-metal_header-generator.c++]: Have another loop for create_max_defines()

File: metal_header/device.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    /* The "compatable" string for the device */
+    std::string compat_string;
+
+    /* The devicetree */
+    const fdt &dtb;
+
+    Device(std::ostream &os, const fdt &dtb, std::string compat_string);
+
+    virtual void create_machine_macros() = 0;
+    virtual void create_defines() = 0;
```
Comment (bsousi5) [inline metal_header/device.h]: may need another for the max defines, say create_max_defines()

File: metal_header/template_device.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
+/* Copyright 2018 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
```
Comment (bsousi5) [inline metal_header/template_device.h]: I assume this template is for developer to use by cloning the file and edit/add the contents as needed.
Can you add a brief comment on top of the file

**Comment by nategraff-sifive** [comment]
@bsousi5 I'm not quite sure what the `METAL_MAX_XXXX` #defines are doing. I don't see where they're used in the library.

I don't think we need extra virtual methods. If there's any kind of accumulation we need to do over all the devices of a given type, we can do that in the constructor for the device, and then just emit that in the `create_defines()` method.

**Comment by bsousi5** [comment]
> @bsousi5 I'm not quite sure what the `METAL_MAX_XXXX` #defines are doing. I don't see where they're used in the library.
> 
> I don't think we need extra virtual methods. If there's any kind of accumulation we need to do over all the devices of a given type, we can do that in the constructor for the device, and then just emit that in the `create_defines()` method.

See metal.h for METAL_MAX_xxxx usage.
If create_defines() can do it even better. I was not sure how it can be done.

**Comment by nategraff-sifive** [comment]
@bsousi5 Got it.

For now I think the best way is actually to just do it all in `create_defines()`. We're iterating over all the matching nodes there anyway, so it's easy enough to just emit the max value.

**Pull request merged**